File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/NegativeLiteralTransformer.java
package dev.openrs2.deob.ast.transform;

import com.github.javaparser.ast.CompilationUnit;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.expr.UnaryExpr;
import dev.openrs2.deob.ast.util.ExprUtils;
import dev.openrs2.deob.ast.util.NodeUtils;

public final class NegativeLiteralTransformer extends Transformer {
	@Override
	public void transform(CompilationUnit unit) {
		NodeUtils.walk(unit, Node.TreeTraversal.POSTORDER, UnaryExpr.class, expr -> {
			var operand = expr.getExpression();
			if (!ExprUtils.isIntegerOrLongLiteral(operand)) {
				return;
			}

			var op = expr.getOperator();
			if (op == UnaryExpr.Operator.PLUS) {
				expr.replace(operand);
			} else if (op == UnaryExpr.Operator.MINUS) {
				expr.replace(ExprUtils.negate(operand));
			}
		});
	}
}

File: deob-ast/src/main/java/dev/openrs2/deob/ast/transform/NegativeLiteralTransformer.kt
package dev.openrs2.deob.ast.transform

import com.github.javaparser.ast.CompilationUnit
import com.github.javaparser.ast.expr.UnaryExpr
import dev.openrs2.deob.ast.util.isIntegerOrLongLiteral
import dev.openrs2.deob.ast.util.negate
import dev.openrs2.deob.ast.util.walk
import javax.inject.Singleton

@Singleton
class NegativeLiteralTransformer : Transformer() {
    override fun transformUnit(
        units: Map<String, CompilationUnit>,
        unit: CompilationUnit
    ) {
        unit.walk { expr: UnaryExpr ->
            val operand = expr.expression
            if (!operand.isIntegerOrLongLiteral()) {
                return@walk
            }

            @Suppress("NON_EXHAUSTIVE_WHEN")
            when (expr.operator) {
                UnaryExpr.Operator.PLUS -> expr.replace(operand)
                UnaryExpr.Operator.MINUS -> expr.replace(operand.negate())
            }
        }
    }
}
